_posts/2024-04-30-search-indexing-rollover.adoc

```diff
@@ -0,0 +1,281 @@
+Look no further. In this post, we'll cover how you can approach this problem
+and solve it in practice with a few low-level APIs provided by Hibernate Search.
+
+The approach suggested in this post is based on the fact that Hibernate Search is using
+link:{hibernate-search-docs-url}#backend-elasticsearch-indexlayout[aliased indexes],
+and communicates with the actual index through a read/write alias, depending on the operation it needs to perform.
+For example, a search operation will be routed to a read index alias,
+while the indexing operation will be sent to a write index alias.
+
+image::initial-app.png[]
+
+NOTE: This approach is implemented and successfully used in our Quarkus application that backs the guides'
+search of https://quarkus.io/guides/[quarkus.io/guides/].
+You can see the complete implementation here:
+link:https://github.com/quarkusio/search.quarkus.io/blob/d956b6a1341d8693fa1d6b7881f3840f48bdaacd/src/main/java/io/quarkus/search/app/indexing/Rollover.java#L44-L331[rollover implementation]
+and link:https://github.com/quarkusio/search.quarkus.io/blob/d956b6a1341d8693fa1d6b7881f3840f48bdaacd/src/main/java/io/quarkus/search/app/indexing/IndexingService.java#L226-L244[rollover usage].
+
+Applications using Hibernate Search can keep their search indexes up-to-date by updating the index gradually,
+as the data on which the index documents are based is modified, providing a near real-time index synchronisation.
+On the other hand, if the search requirements allow for a delay in synchronisation
+or the data is updated only at certain times of day, the option of mass indexing can effectively keep the indexes up-to-date.
+The link:{hibernate-search-docs-url}[Hibernate Search documentation] provides more information about these approaches
+and other Hibernate Search capabilities.
+
+The application discussed in this post is using the mass indexing approach.
+This means that at certain events, e.g., when a new version of the application is deployed or a scheduled time is reached,
+the application has to process the documentation guides and create search index documents from them.
+
+Now, since we would want to keep our application providing results to any search operations and add/update documents to the indexes
+we cannot perform a simple reindex operation (purge all documents within an index and mass-index them back)
+using a link:{hibernate-search-docs-url}#search-batchindex-massindexer[mass indexer],
+or the recently added link:{quarkus-hibernate-search-docs-url}#management[management Quarkus endpoint],
+as this will drop all existing documents from the index, and search operations will not be able to match them anymore.
+
+Instead, we can create a new index with the same schema and route any write operations to it.
+
+image::write-app.png[]
+
+Since, at the moment, Hibernate Search does not provide the rollover feature out of the box
```

Comment:
```suggestion
Since Hibernate Search does not provide the rollover feature out of the box (https://hibernate.atlassian.net/browse/HSEARCH-3499[yet])
```